include/boost/fusion/container/vector/vector.hpp

```diff
@@ -9,12 +9,620 @@
+                  std::size_t
+                , fusion::result_of::size<typename remove_reference<Sequence>::type>::value
+              >
+        {};
+
+        // Object proxy since preserve object order
+        template <std::size_t, typename T>
+        struct store
+        {
+            BOOST_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            store()
+                : elem()
+            {}
+
+            BOOST_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            store(store const& rhs)
+                : elem(rhs.get())
+            {}
+
+            //BOOST_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            //store(T const& val)
+            //    : elem(val)
+            //{}
+
+            BOOST_CXX14_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            store&
+            operator=(store const& rhs)
+            {
+                elem = rhs.get();
+                return *this;
+            }
+
+            BOOST_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            store(store&& rhs)
+                : elem(std::forward<T>(rhs.get()))
+            {}
+
+            BOOST_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            store(typename remove_reference<T>::type&& val)
+                : elem(std::forward<T>(val))
+            {}
+
+            BOOST_CXX14_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            store&
+            operator=(store&& rhs)
+            {
+                elem = std::forward<T>(rhs.get());
+                return *this;
+            }
+
+            template <typename U>
+            BOOST_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            store(U&& rhs
+                , typename enable_if<
+                      mpl::or_<
+                          is_convertible<typename traits::remove_rvalue_reference<U>::type, T>
+                        , detail::is_constructible<T, typename traits::remove_rvalue_reference<U>::type>
+                      >
+                  >::type* = 0)
+                : elem(std::forward<U>(rhs))
+            {}
+
+            template <typename U>
+            BOOST_CXX14_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            store&
+            operator=(U&& rhs)
+            {
+                elem = std::forward<U>(rhs);
+                return *this;
+            }
+
+            BOOST_CXX14_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            T      & get()       { return elem; }
+            BOOST_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            T const& get() const { return elem; }
+
+            T elem;
+        };
+
+        template <std::size_t I, typename... T>
+        struct data;
+
+        template <std::size_t I>
+        struct data<I>
+        {
+            typedef mpl::int_<(int)I> size;
+
+            BOOST_CXX14_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            void
+            assign() BOOST_NOEXCEPT
+            {}
+        };
+
+        template <std::size_t I, typename T, typename... Tail>
+        struct data<I, T, Tail...>
+            : store<I, T>, data<I + 1, Tail...>
```

Comment:
I'm not sure how difficult this would be to change, but you might want to consider multi-inheritance expansion instead of recursive tail expansion inheritance.

About a month ago on IRC, K-ballo pointed out [this blog post](http://www.elbeno.com/blog/?p=1104) incorrectly described the "lambda trick" mentioned on the boost spirit mailing list. The `std::tuple` code in the blog post was slower to compile than the lambda version because of the libstdc++ implementation of `std::tuple`. When compiling the same code with `libc++`, I got the `std::tuple` version to to compile faster than the lambda version (clang 3.5 in c++11 mode on OSX).

libc++ expands the tuple using multiple inheritance, and libstdc++ uses recursive inheritance (similar to this implementation). When using multiple inheritance, larger tuples don't hit the recursive template expansion limit, since the expansion occurs at the same depth level. This instantation difference (and possibly other factors) enable it to compile much faster. You'd have to use an integer sequence that gets unpacked at the same time, but it should be doable. I'm not sure how difficult this is to change in this implementation (I don't expect it being too difficult), but perhaps this is something to profile/change later. Actually, I'll try to make the changes and compare the differences tomorrow, regardless of the status of this patch.